### PR TITLE
Final overhaul of list membership and associated relations

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
It's obviously an afternoon for big PRs! This finally organises list membership in a sane, logical and easy-to-use way. As a happy consequence this solves issues #70, #155 and #220.

The major changes are:
- Moved `Data/List.Any.BagAndSetEquality` to `Data.List.Relation.BagAndSetEquality`
- Moved the sublist relation in `Data.List.Membership` into it's own modules `Data.List.Relation.Sublist.(Propositional/Setoid)`.
- Membership is now in four modules in `Data.List.Membership` directory:
  - `Membership.Setoid` parameterised by a `Setoid`
  - `Membership.DecSetoid` parameterised by a `DecSetoid`
  - `Membership.Propositional` implictly parameterised by a type
  - `Membership.DecPropositional` parameterised by a proof of decidable equivalence for an implicit type.
- The proofs concerning the membership operations `find` and `lose` and are now in the correct membership property files instead of being confusingly misplaced in `Data.Any.Properties`.
- Added a whole bunch more proofs about membership
 
This structure makes it easy to simply pick the type of membership you're interested in and get all the misfix notation working automatically upon module import at the top of the file.

I should also mention that I chose to prefix the introduction and elimination membership proofs with `∈` as we may at some point want to have corresponding proofs with `∉`.

### CODE REVIEW

File: src/Data/List/Membership/Setoid/Properties.agda [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+------------------------------------------------------------------------
+-- filter
+
+module _ {c ℓ p} (S : Setoid c ℓ) {P : Pred (Carrier S) p} where
+
+  open Setoid S using (_≈_; sym)
+  open Membership S using (_∈_)
+
+  ∈-filter⁺ : (P? : Decidable P) → P Respects _≈_ →
```
Comment (gallais) [inline src/Data/List/Membership/Setoid/Properties.agda]: These two parameters `(P? : Decidable P)` and `P Respects _≈_` are common to both
`∈-filter⁺` and `∈-filter⁻` and could be pushed in the anonymous module.

File: src/Data/List/Membership/Setoid/Properties.agda [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+  ... | yes _ = there (∈-filter⁺ P? resp v∈xs Pv)
+  ... | no  _ = ∈-filter⁺ P? resp v∈xs Pv
+
+  ∈-filter⁻ : (P? : Decidable P) → P Respects _≈_ →
+              ∀ {v xs} → v ∈ filter P? xs → v ∈ xs × P v
+  ∈-filter⁻ P? resp {v} {[]}     ()
+  ∈-filter⁻ P? resp {v} {x ∷ xs} v∈f[x∷xs] with P? x
+  ... | no  _  = Prod.map there id (∈-filter⁻ P? resp v∈f[x∷xs])
+  ... | yes Px with v∈f[x∷xs]
+  ...   | here  v≈x     = here v≈x , resp (sym v≈x) Px
```
Comment (gallais) [inline src/Data/List/Membership/Setoid/Properties.agda]: `=` out of alignment

File: src/Data/List/Membership/Setoid/Properties.agda [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+module _ {c ℓ} (S : Setoid c ℓ) where
+
+  open Membership S using (_∈_)
+
+  ∈-length : ∀ {x xs} → x ∈ xs → 1 ≤ length xs
+  ∈-length {_} {_ ∷ xs} (here px)    = s≤s z≤n
+  ∈-length {_} {_ ∷ xs} (there x∈xs) =
+    ≤-trans (∈-length x∈xs) (n≤1+n (length xs))
+
```
Comment (gallais) [inline src/Data/List/Membership/Setoid/Properties.agda]: Can't we have `n≤1+n (length _)` and drop the pattern matching on implicit arguments?

File: src/Data/Product/Properties.agda [hunk 4]
```diff
@@ -18,3 +18,11 @@ module _ {a b} {A : Set a} {B : A → Set b} where
...
  ,-injectiveʳ refl = refl
 
+ proj₁-injective : ∀ {a} {b : B a} {c : Σ A B} → c ≡ (a , b) → proj₁ c ≡ a
+ proj₁-injective refl = refl
+
+ proj₂-injective : ∀ {a} {b : B a} {c} → c ≡ (a , b) → proj₂ c ≡ b
+ proj₂-injective refl = refl
+
+ proj-injective : ∀ (ab : Σ A B) → ab ≡ (proj₁ ab , proj₂ ab)
+ proj-injective (_ , _) = refl
```
Comment (gallais) [inline src/Data/Product/Properties.agda]: Isn't this true by eta-equality (i.e. no need to pattern-match)?

File: src/Data/List/All/Properties.agda [hunk 5]
```diff
@@ -258,7 +260,7 @@ module _ {a p} {A : Set a} {P : A → Set p} where
 ------------------------------------------------------------------------
 -- tabulate
 
-module _ {a p} {A : Set a} {P : A → Set p} where
+module _ {a p} {A : Set a} {P : A → Set p}  where
```
[the remark this answers is not in the document]
Comment (MatthewDaggitt) [reply src/Data/List/All/Properties.agda]: Fixed.

File: src/Data/List/Membership/Setoid/Properties.agda (hunk 1 above)
Comment (MatthewDaggitt) [reply src/Data/List/Membership/Setoid/Properties.agda]: Done

File: src/Data/List/Membership/Setoid/Properties.agda (hunk 2 above)
Comment (MatthewDaggitt) [reply src/Data/List/Membership/Setoid/Properties.agda]: Fixed

File: src/Data/List/Membership/Setoid/Properties.agda (hunk 3 above)
Comment (MatthewDaggitt) [reply src/Data/List/Membership/Setoid/Properties.agda]: Nope but we can have `n≤1+n _`!

File: src/Data/Product/Properties.agda (hunk 4 above)
Comment (MatthewDaggitt) [reply src/Data/Product/Properties.agda]: Yup, you're right. Removed.

File: src/Data/List/Membership/Propositional/Properties.agda [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+
+------------------------------------------------------------------------
+-- Other properties
+
+-- Only a finite number of distinct elements can be members of a
+-- given list.
+
+finite : ∀ {a} {A : Set a} (f : ℕ ↣ A) →
+         ∀ xs → ¬ (∀ i → Inj.Injection.to f ⟨$⟩ i ∈ xs)
+finite         inj []       ∈[]   with ∈[] zero
```
[the remark this answers is not in the document]
Comment (MatthewDaggitt) [reply src/Data/List/Membership/Propositional/Properties.agda]: A little tidier now.

**Pull request merged**